3p-ecosystem/src/main/java/com/google/homesampleapp/screens/thread/ThreadFragment.kt

```diff
@@ -130,66 +200,703 @@ class ThreadFragment : Fragment() {
+    // Note that Fragments must call registerForActivityResult() before they are created
     threadClientLauncher =
-        registerForActivityResult(ActivityResultContracts.StartIntentSenderForResult()) { result ->
-          if (result.resultCode == RESULT_OK) {
-            val threadNetworkCredentials =
-                ThreadNetworkCredentials.fromIntentSenderResultData(result.data!!)
-            viewModel.threadPreferredCredentialsOperationalDataset.postValue(
-                threadNetworkCredentials)
-          } else {
-            val error = "User denied request."
-            Timber.d(error)
-            updateThreadInfo(null, "")
+      registerForActivityResult(ActivityResultContracts.StartIntentSenderForResult()) { result ->
+        if (result.resultCode == RESULT_OK) {
+          val threadNetworkCredentials =
+            ThreadNetworkCredentials.fromIntentSenderResultData(result.data!!)
+          viewModel.setThreadCredentialsInfo(null, threadNetworkCredentials)
+        } else {
+          val error = "User denied request."
+          Timber.d(error)
+          viewModel.setThreadCredentialsInfo(null, null)
+        }
+      }
+  }
+
+  // -----------------------------------------------------------------------------------------------
+  // Top level Composables
+
+  @Composable
+  private fun ThreadRoute() {
+    // UI Logic implemented in ThreadNetworkUiState requires Activity.
+    val activity = LocalContext.current.getActivity()
+
+    // All Thread Network actions are handled by UI State class ThreadNetworkUiState.
+    // We are not using a ViewModel because ThreadNetwork actions require Activity
+    // and the lifetime of a ViewModel is longer than Activity.
+    val threadNetworkUiState = remember(activity) { ThreadNetworkUiState(activity!!, viewModel) }
+
+    // All calls into the ThreadNetworkUiState are encapsulated within this lambda which makes
+    // it possible to avoid exposing ThreadNetworkUiState into the composable ThreadScreen
+    // (which is less optimal for recompositions as there is a lot of stuff in ThreadNetworkUiState).
+    // ActionRequest provides all the information needed by ThreadNetworkUiState
+    // to properly process the Action triggered by the user.
+    //
+    // Note that this makes it challenging to handle UIState functions that are suspendable.
+    // At the point of invocation in the composable, one would normally use rememberCoroutineScope
+    // but the call must be done directly on the UIState object, which we strive to avoid.
+    // For now, suspend functions are handled in the ViewModel by using viewModelScope.
+    // [Not possible to have a lambda for a suspend function that needs arguments:
+    // https://youtrack.jetbrains.com/issue/KT-51067/Function-for-creating-suspending-lambdas-doesnt-allow-lambda-parameters
+    // https://medium.com/livefront/suspending-lambdas-in-kotlin-7319d2d7092a]
+    val onThreadNetworkAction =
```

Comment:
This val should also be remembered with the `threadNetworkUiState` as a key:

```
val onThreadNetworkAction = remember(threadNetworkUiState) {
      { actionRequest: ActionRequest ->
        threadNetworkUiState.processAction(actionRequest)
      }
}      
```

Reply:
Done.